src/main/java/io/cryostat/recordings/ArchiveRequestGenerator.java

```diff
@@ -0,0 +1,191 @@
+    public static final String GRAFANA_ARCHIVE_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.GrafanaArchiveUploadRequest";
+    public static final String GRAFANA_ACTIVE_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.GrafanaActiveUploadRequest";
+    private static final String ARCHIVE_RECORDING_SUCCESS = "ArchiveRecordingSuccess";
+    private static final String ARCHIVE_RECORDING_FAIL = "ArchiveRecordingFailed";
+    private static final String GRAFANA_UPLOAD_SUCCESS = "GrafanaUploadSuccess";
+    private static final String GRAFANA_UPLOAD_FAIL = "GrafanaUploadFailed";
+    private final ExecutorService executor;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Inject private EventBus bus;
+    @Inject private RecordingHelper recordingHelper;
+
+    @ConfigProperty(name = ConfigProperties.CONNECTIONS_FAILED_TIMEOUT)
+    Duration timeout;
+
+    public ArchiveRequestGenerator(ExecutorService executor) {
+        this.executor = executor;
+    }
+
+    public Future<String> performArchive(ArchiveRequest request) {
+        Objects.requireNonNull(request.getRecording());
+        return executor.submit(
+                () -> {
+                    logger.info("Job ID: " + request.getId() + " submitted.");
+                    try {
+                        recordingHelper.archiveRecording(request.getRecording(), null, null).name();
+                        logger.info("Recording archived, firing notification");
+                        bus.publish(
+                                MessagingServer.class.getName(),
+                                new Notification(
+                                        ARCHIVE_RECORDING_SUCCESS,
+                                        Map.of("jobId", request.getId())));
+                        return request.getId();
+                    } catch (Exception e) {
+                        logger.info("Archiving failed");
+                        bus.publish(
+                                MessagingServer.class.getName(),
+                                new Notification(
+                                        ARCHIVE_RECORDING_FAIL, Map.of("jobId", request.getId())));
+                        throw new CompletionException(e);
+                    }
+                });
+    }
+
+    @ConsumeEvent(value = ARCHIVE_ADDRESS)
+    public void onMessage(ArchiveRequest request) {
+        try {
+            performArchive(request);
+        } catch (Exception e) {
+            logger.warn("Exception thrown while servicing request: ", e);
+        }
+    }
+
+    @ConsumeEvent(value = GRAFANA_ARCHIVE_ADDRESS)
+    public void onMessage(GrafanaArchiveUploadRequest request) {
+        try {
+            logger.info("Job ID: " + request.getId() + " submitted.");
+            String result =
+                    recordingHelper
+                            .uploadToJFRDatasource(request.getPair())
+                            .await()
```

Comment:
This `.await()` blocks execution of the current thread, so this event handler should also have `blocking = true`.